contracts/VolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,152 @@
+pragma solidity ^0.4.18;
+
+import "./ERC20Interface.sol";
+import "./PermissionGroups.sol";
+
+interface VolumeImbalanceRecorderInterface {
+  function getImbalance( ERC20 token, uint priceUpdateBlock, uint currentBlock)
+    public view returns(int totalImbalance, int currentBlockImbalance);
+}
+
+contract VolumeImbalanceRecorder is PermissionGroups, VolumeImbalanceRecorderInterface {
+
+  uint public constant SLIDING_WINDOW_SIZE = 10;
+
+  struct TokenControlInfo {
+    uint minimalRecordResolution; // can be roughly 1 cent
+    uint maxPerBlockImbalance; // in twei resolution
+    uint maxTotalImbalance; // max total imbalance (without price updates)
+                            // before halting trade
+  }
+
+  mapping(address => TokenControlInfo) tokenControlInfo;
+
+  struct TokenImbalanceData {
+    int64  lastBlockBuyUnitsImbalance;
+    uint64 lastBlock;
+
+    int64  totalBuyUnitsImbalance;
+    uint64 lastPriceUpdateBlock;
+  }
+
+  mapping(address => mapping(uint=>TokenImbalanceData)) tokenImbalanceData;
+
+  event SetTokenControlInfo( ERC20 token,
+                             uint minimalRecordResolution,
+                             uint maxPerBlockImbalance,
+                             uint maxTotalImbalance );
+
+  function setTokenControlInfo( ERC20 token,
+                                uint minimalRecordResolution,
+                                uint maxPerBlockImbalance,
+                                uint maxTotalImbalance ) public onlyAdmin {
+    tokenControlInfo[token] = TokenControlInfo(
+                                minimalRecordResolution,
+                                maxPerBlockImbalance,
+                                maxTotalImbalance);
+    SetTokenControlInfo( token,
+                         minimalRecordResolution,
+                         maxPerBlockImbalance,
+                         maxTotalImbalance );
+  }
+
+  function getTokenControlInfo( ERC20 token ) public view returns(uint,uint,uint) {
+    return (tokenControlInfo[token].minimalRecordResolution,
+            tokenControlInfo[token].maxPerBlockImbalance,
+            tokenControlInfo[token].maxTotalImbalance);
+  }
+
+  function getImbalanceInRange( ERC20 token, uint startBlock, uint endBlock ) public view returns(int buyImbalance ) {
```

Comment:
consider making this internal - so only getImbalance is used. or some other function that multiply resolution. 
for the user its not clear enough this is should be mul with resolution.